File: 3SumClosest.java
/**
 * LeetCode 16 3Sum Closest
 */
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        if (nums == null || nums.length < 3) {
            return 0;
        }
        Arrays.sort(nums);
        int res = nums[0] + nums[1] + nums[2];
        for (int i = 0; i < nums.length - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            int pf = i + 1;
            int pe = nums.length - 1;
            while (pf < pe) {
                int cur = nums[pf] + nums[pe] + nums[i];
                res = Math.abs(target - cur) < Math.abs(target - res) ? cur : res;
                if (cur > target) {
                    pe--;
                } else if (cur < target) {
                    pf++;
                } else {
                    return res;
                }
            }
        }
        return res;
    }
}

File: 3SumClosest.go

func threeSumClosest(nums []int, target int) int {
	sort.Ints(nums)
	n := len(nums)
	best := math.MaxInt32

	update := func (cur int) {
		if abs(cur - target) < abs(best - target) {
			best = cur
		}
	}

	for i := 0;i < n;i++ {
		if i > 0 && nums[i] == nums[i - 1] {
			continue
		}
		j,k := i + 1,n - 1
		for j < k {
			sum := nums[i] + nums[j] + nums[k]
			if sum == target {
				return target
			}
			update(sum)
			if sum > target {
				k0 := k - 1
				for j < k0 && nums[k0] == nums[k] {
					k0--
				}
				k = k0
			} else {
				j0 := j + 1
				for j0 < k && nums[j0] == nums[j] {
					j0++
				}
				j = j0
			}
		}
	}
	return best
}

func abs(v int) int {
	if v < 0 {
		return -v
	}
	return v
}
